OpenMEEG/include/analytics.h

```diff
@@ -162,56 +169,116 @@ namespace OpenMEEG {
+    // This just solves with a least squares approach the linear system
+    //                 r = l*r0+m*r1+(1-l-m)*r2
+    // Least squares is needed because if we select a subset of equations, we may run in a
+    // singular case. This leads to the 2x2 system:
+    //
+    //              |r0-r2]^2       * l + (r0-r2).(r1-r1) * m = (r0-r2).(r-r2)
+    //              (r0-r2).(r1-r1) * l + |r1-r2]^2       * m = (r1-r2).(r-r2)
+    //
+    //  which is solved using the Kramer formualas.             
+    //
+    // TODO: It is a little bit silly to do this. The points that are passed by the integrator
+    // are computed from their barycentric coordinates and we now do the opposite computation.
+    // A more clever way would be to pass the barycentric coordinates directly and to compute
+    // the point in the analytic functions. This will totally suppress the need for this class.
+
+    class OPENMEEG_EXPORT BarycentricCoordinates {
     public:
 
-        analyticDipPotDer(const Dipole& dip,const Triangle& T): dipole(dip) {
+        BarycentricCoordinates(const Triangle& T) {
 
             const Vect3& p0 = T.vertex(0);
             const Vect3& p1 = T.vertex(1);
             const Vect3& p2 = T.vertex(2);
 
-            const Vect3& p1p0 = p0-p1;
-            const Vect3& p2p1 = p1-p2;
-            const Vect3& p0p2 = p2-p0;
-            const Vect3& p1p0n = p1p0/p1p0.norm();
-            const Vect3& p2p1n = p2p1/p2p1.norm();
-            const Vect3& p0p2n = p0p2/p0p2.norm();
-
-            const Vect3& p1H0 = dotprod(p1p0,p2p1n)*p2p1n;
-            H0 = p1H0+p1;
-            H0p0DivNorm2 = p0-H0;
-            H0p0DivNorm2 = H0p0DivNorm2/H0p0DivNorm2.norm2();
-            const Vect3& p2H1 = dotprod(p2p1,p0p2n)*p0p2n;
-            H1 = p2H1+p2;
-            H1p1DivNorm2 = p1-H1;
-            H1p1DivNorm2 = H1p1DivNorm2/H1p1DivNorm2.norm2();
-            const Vect3& p0H2 = dotprod(p0p2,p1p0n)*p1p0n;
-            H2 = p0H2+p0;
-            H2p2DivNorm2 = p2-H2;
-            H2p2DivNorm2 = H2p2DivNorm2/H2p2DivNorm2.norm2();
-
-            n = -crossprod(p1p0,p0p2);
-            n.normalize();
+            r20 = p0-p2;
+            r21 = p1-p2;
+            r2  = p2;
+
+            const double norm_r20_squared = r20.norm2();
+            const double norm_r21_squared = r21.norm2();
+            const double prod_r20_r21     = dotprod(r20,r21);
+
+            const double det = norm_r20_squared*norm_r21_squared-sqr(prod_r20_r21);
+            const double invdet = 1.0/det;
+
+            c11 = norm_r20_squared*invdet;
+            c22 = norm_r21_squared*invdet;
+            c12 = prod_r20_r21*invdet;
+        }
+
+        Vect3 operator()(const Vect3& r) const {
+
+            const Vect3& u = r-r2;
+            const double b1 = dotprod(r20,u);
+            const double b2 = dotprod(r21,u);
+            const double l = b1*c22-b2*c12;
+            const double m = b2*c11-b1*c12;
+            return Vect3(l,m,1-l-m);
         }
```

Comment:
I changed all of this code, because honestly I was not able to understand what it did from the formula. And there was no comment to explain it. The new version is explained and costs less (both in space, which is not very crucial, but also in time). I verified that it gave strictly the same results. A further improvement would be to completely eliminate this calculations. Indeed, we compute the points using barycentric coordinates and then pass the points to these functions... to re-compute the barycentric coordinates. It would be simpler to pass directly the barycentric coodinates and to recompute the point in the analytic functions. I will do this in another PR.